### ISSUE TITLE
Level Access Issue 2: Help Users (esp. screen reader users) Identify Links to the Answer Key

### ISSUE THREAD

**Issue by openstaxalina** [openstax/cnx-recipes#5256]
# Context
When we link to an answer in the end-of-book Answer Key, it is a subtle link on the question number (see the odd-numbered exercises in the screenshot below from [this book section](https://openstax.org/books/elementary-algebra-2e/pages/1-1-introduction-to-whole-numbers) )

<img width="506" alt="Screenshot 2023-09-27 at 5 15 12 PM" src="https://github.com/openstax/cnx-recipes/assets/7863733/432a3178-b713-4e7c-8440-4b54b774d892">

# New specifications or link to documentation cards
I've copied the full recommendation from Level Access below. My takeaways are:
- this is way too many links to manually adjust in the content
- is there any kind of automation we can apply, when it's a link to the Answer Key, to basically say "this is a link to the solution of problem [#]"? 

<p>Provide an accessible name of each link that accurately reflects the target and purpose of the link when the link and its surrounding text is taken into context. This can be done by changing text that is used to calculate a link's accessible name, or by modifying text before the link, such as in the same sentence, paragraph, list item, etc. Even when combined with text in context, it may still be helpful to users of screen readers to provide an accessible link name that makes the link's purpose clear out of context.</p>
<p>Link text that is too long, too short, or uses common terms or duplicates other link text on the same page needs to be reviewed. to determine if it accurately communicates the link purpose in context.</p>
<p>If the accessible name of the link cannot be updated, authors should ensure that the content surrounding the link aids in identifying the links purpose. Acceptable methods of providing purpose via link context includes indicating the purpose in the same:</p>
<ol>
<li>Sentence, paragraph, or list item or</li>
<li>Table cell as the link or</li>
<li>In the table header cell for a link in a data table</li>
</ol>
<h3>Accessible name calculation</h3> 
<p>The accessible name of a link(https://www.w3.org/TR/html-aam-1.0/#a-element-with-a-href-attribute) is calculated based on the name calculation algorithm (https://www.w3.org/TR/accname-1.1/%22%3Ename). In simplified terms, the following is used to determine the name and takes precedence rather than concatenating each method together.</p>
<ul>
<li> <a href="[http://www.w3.org/TR/2012/NOTE-WCAG20-TECHS-20120103/ARIA1"](http://www.w3.org/TR/2012/NOTE-WCAG20-TECHS-20120103/ARIA1%22)> <code>aria-labelledby</code> </a> referenced content</li>
<li> <code>aria-label</code> </li>
<li>element subtree including images with alternative text, and CSS off-screen positioned text</li>
<li>the <code>title</code> attribute</li>
</ul>
It is important to remember that overly verbose link names may be difficult to understand and when located in data table header cells may create extraneous verbosity when users of screen reader navigate associated data table cells.
<h3>Accessible Descriptions</h3>
An accessible description can be provided by the <code>title</code> attribute or aria-describedby can be used to supplement the accessible name. However, due to how screen readers announce the accessible description, it is recommended to communicate the link purpose via the link's accessible name and/or the link's context.
Note: The title attribute only acts as accessible name when there is no other method of calculating a name, otherwise it becomes the accessible description of the link.
<p>When the link has an accessible name, such as one provided by text content, an image with alternative, or a label provided by ARIA, the <code>title</code> attribute acts as an accessible description. This description may not be announced by certain navigation methods, by default, by many major screen readers, like JAWS, VoiceOver or NVDA. It may be announced when tabbing or shift+tabbing to the link but is not announced when arrowing to the link.
<p>For users who are not using a screen reader, the <code>title</code> is also displayed when the mouse is moved over a link; however, this feature is not keyboard accessible in nearly all user agents. This is one reason why only supplementary information can be provided via this method.</p>
<h3>Links to file formats or external sites</h3>
<p>For links that indicate the type of media or an external destination, an image can also be inserted inside the link with appropriate <code>alt</code> text indicating the context (e.g., using a PDF icon to differentiate a PDF link from an HTML link). This approach works well for links that direct users to other sites or open new windows. In these cases, an icon image with appropriate <code>alt</code>text can be used to indicate the actions. WCAG 2 Level A/AA does not require mention that links open windows or point to PDF files but does indicate it as advisory. For this best practice, the link text in context for accessibility only need communicate what is seen visually -- if a link's purpose is ambiguous to everyone, it is not a failure of this test.</p>
<h3>Links that are repurposed for UI elements</h3>
<p>When a link combined with a role represents a non-link UI element, such as a page tab, refer to the tests related to those controls for guidance.</p>
<h3>A Note about Off-screen positioned text</h3>
<p>Text that is positioned off-screen using CSS can be a good solution for users of screen readers; however, this text is not accessible to anyone who has CSS turned on and is not using a screen reader. For this reason, it is recommended that this method only be used with this supplementary information is only needed by users of screen readers or where it is duplicated using other methods such as the via text in the same sentence, paragraph, list items, table cell, etc. Typically, this method works by creating a CSS style, such as <code>.sronly</code> which places the text content off-screen using CSS and then this class is applied to a <code>span</code> element within the anchor that contains the text to hide.

## Impact 

# Acceptance Criteria

# Content
```
<exercise id="fs-id1170655190140">
<problem id="fs-id1170655190142">
<para id="fs-id1170655190144">51,493
<span class="token">ⓐ</span> 1,
<span class="token">ⓑ</span> 4,
<span class="token">ⓒ</span> 9,
<span class="token">ⓓ</span> 5,
<span class="token">ⓔ</span> 3</para>
</problem>
<solution id="fs-id1170655205414">
<para id="fs-id1170655205416"><span class="token">ⓐ</span> thousands <span class="token">ⓑ</span> hundreds <span class="token">ⓒ</span> tens <span class="token">ⓓ</span> ten thousands <span class="token">ⓔ</span> ones</para>
</solution>
</exercise>
```

## Location(s)
- **osbooks repo:** multiple, but these examples are from github.com/openstax/osbooks-prealgebra-bundle
- **collection slug:** elementary-algebra-2e
- **ref/branch:** main
- **chapter/section:** 1.1
- **surrounding text:** "In the following exercises, find the place value of each digit in the given numbers."

**Comment by openstaxalina** [comment]
Discussed with @kswanson33 and @therealmarv -- we think that the markup is consistent enough for some kind of automation to be applied when something IS a link and IS a link to the Answer Key. There is a step in cookbook where the solutions are separated out from the rest of the content, and we would likely add something (aria label? tool tip indicator?) there.

**Comment by openstaxalina** [comment]
I will discuss with @erikayvonne and @lwtchu to confirm whether we want the content to show to all users or only screen readers users.

**Comment by openstaxalina** [comment]
Customer Service hasn't received feedback: https://openstax.slack.com/archives/C0V220ZRN/p1696519252094859?thread_ts=1696519056.127149&cid=C0V220ZRN

**Comment by openstaxalina** [comment]
[Full response from Level Access has been logged here](https://openstax.atlassian.net/wiki/spaces/UNIFIED/pages/2311290881/Level+Access+support+tickets) (search `Links to/from Answer Key`)

tl;dr -- adding tooltips might be useful, but will not resolve the specific a11y bug logged here. Therefore, I think we should use this card to implement an aria label on links to AND from the Answer Key. The label should specify the question number. For example:

- "answer to exercise [#]" (when going TO the Answer Key)
- "exercise [#]" (when coming FROM the Answer Key)

@lwtchu please let me know if you team would require any further info to move forward with this card within the next few sprints. We can of course discuss at any upcoming calls as needed!

**Comment by openstaxalina** [comment]
Here's what Level Access specifically said about adding aria labels (again the **full** response is linked above):

aria-labels serve as accessible names (for links and buttons) for screen reader users, allowing assistive technologies to convey the link's purpose effectively. Unlike tooltips, aria-labels directly modify the accessible name of an element, ensuring that screen readers announce a meaningful description of the link. For users with disabilities, including those who are blind or have low vision, aria-labels are essential for understanding the link's purpose without relying on visual cues. These labels should offer a more descriptive and meaningful name for the link, addressing the accessibility needs of users who rely on screen readers. For example, you can use aria-label="Chapter 1" to provide a clear accessible name for screen readers.

**Comment by openstaxalina** [comment]
@philschatz and @lwtchu pointed out that it could be confusing if we add the aria label as-specified, since the number is serving two purposes:

- it's a link to/from the answer (which is why we want to add an aria label)
- it's providing the number of the exercise

I will get back to Level Access with this question. If there are any other thoughts/options/questions I should ask, please let me know today.

Also, we have an upcoming call with Level Access on 10/31 (👻) and could request this be a topic and more folks could join. We might want to do that if this next email exchange doesn't resolve our concerns. cc @kswanson33 @LostMojo @jbwilson8 @erikayvonne

**Comment by erikayvonne** [comment]
@openstaxalina I'm not sure we'd get help at our monthly check-in calls since those folks aren't technical - but I know @Brooke-Fail was planning to use one of our live consultations to discuss screenreader testing & keyboard navigation (and open issues we have related to those). Maybe it would make sense to roll this into that conversation, if we y'all end up needing to discuss further?

**Comment by openstaxalina** [comment]
I had dropped the ball on this (😬) but just emailed Level Access with our [question of how to handle the dual purposes of the number/link](https://github.com/openstax/cnx-recipes/issues/5256#issuecomment-1759803408).
